code/modules/antagonists/mind_flayer/flayer_power.dm

```diff
@@ -0,0 +1,277 @@
+	return isspell(spell)
+
+
+/*Given a spell, checks if a mindflayer is able to afford, and has the prerequisites for that spell.
+* If so it adds the ability and increments the category stage if needed, then returns TRUE
+* otherwise, returns FALSE
+*/
+
+/datum/antagonist/mindflayer/proc/try_purchase_spell(datum/spell/flayer/to_add)
+	var/datum/spell/flayer/existing_spell = has_spell(to_add)
+	if(existing_spell)
+		return try_upgrade_spell(existing_spell)
+
+	if(to_add.current_cost > get_swarms())
+		send_swarm_message("We need [to_add.current_cost - get_swarms()] more swarms for this...")
+		return FALSE
+
+	if(category_stage[to_add.category] < to_add.stage)
+		send_swarm_message("We do not have all the knowledge needed for this.")
+		return FALSE
+	if(to_add.stage == CAPSTONE_STAGE)
+		if(!can_pick_capstone)
+			send_swarm_message("We have already forsaken that knowledge.")
+			return FALSE
+		can_pick_capstone = FALSE
+		send_swarm_message("We evolve to the ultimate being.")
+	if(category_stage[to_add.category] == to_add.stage)
+		category_stage[to_add.category] += 1
+
+	to_add.current_cost = to_add.base_cost
+	adjust_swarms(-to_add.current_cost)
+	add_ability(to_add, src) //Level gets set to 1 when AddSpell is called later, it also handles the cost
+	return TRUE // The reason we do this is cause we don't have the spell object that will get added to the mindflayer yet
+/*
+* Given a spell, checks to see if the mindflayer has all the prerequisites to upgrade a spell. Returns FALSE if the spell was not upgrade, TRUE if it was.
+*/
+/datum/antagonist/mindflayer/proc/try_upgrade_spell(datum/spell/flayer/to_upgrade)
+	if(to_upgrade.level >= to_upgrade.max_level)
+		send_swarm_message("That function is already at it's strongest.")
+		return FALSE
+	if(to_upgrade.current_cost > get_swarms())
+		send_swarm_message("We need [to_upgrade.current_cost - get_swarms()] more swarms for this...")
+		return FALSE
+	adjust_swarms(-to_upgrade.current_cost)
+	to_upgrade.level += 1
+	to_upgrade.current_cost = to_upgrade.static_upgrade_increase ? (to_upgrade.static_upgrade_increase + to_upgrade.current_cost) : (to_upgrade.base_cost * (to_upgrade.level + 1))
+	add_ability(to_upgrade, src)
+	return TRUE
+
+/*Given a passive, checks if a mindflayer is able to afford, and has the prerequisites for that spell.
+* If so it adds the ability and increments the category stage if needed, then returns TRUE
+* otherwise, returns FALSE
+*/
+/datum/antagonist/mindflayer/proc/try_purchase_passive(datum/mindflayer_passive/to_add)
+	var/datum/mindflayer_passive/existing_passive = has_passive(to_add)
+	if(existing_passive)
+		if(existing_passive.level >= to_add.max_level)
+			send_swarm_message("That function is already at it's strongest.")
+			return FALSE
+		to_add.current_cost = existing_passive.current_cost
+
+	if(to_add.current_cost > get_swarms())
+		send_swarm_message("We need [to_add.current_cost - get_swarms()] more swarms for this...")
```

Comment:
I don't think I understand what this would do, there shouldn't be any need for an escape character right?

Reply:
This will conditionally add an `s` if the number in the previous thing in brackets needs it. So it'll do like `1 more swarm` or `2 more swarms` properly